### PR TITLE
Index hints

### PR DESCRIPTION
**Pull request by danielhlarkin**

### CODE REVIEW

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3489/

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3491/

File: arangod/Aql/IndexHint.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+namespace aql {
+
+/// @brief container for index hint information
+class IndexHint {
+ public:
+  enum HintType : uint8_t { Illegal, None, Simple };
+
+ public:
+  IndexHint();
+  IndexHint(AstNode const* node);
```
Comment (jsteemann) [inline arangod/Aql/IndexHint.h]: can we make this constructor explicit?

File: arangod/Transaction/Methods.cpp [hunk 2]
```diff
@@ -666,6 +668,36 @@ std::pair<bool, bool> transaction::Methods::findIndexHandleForAndNode(
...
+      if (matched != nullptr) {
+        considerIndex(matched);
+        if (bestIndex != nullptr) {
+          break;
+        }
+      }
+    }
+
+    if (hint.isForced() && bestIndex == nullptr) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_QUERY_FORCED_INDEX_HINT_UNUSABLE);
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: Ideally we could also give a description of the non-found index hint here, e.g. by putting the supplied IndexHint into a human-readable form and putting it into the error message.

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3506/

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3517/

**Pull request merged**